tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/EnhancementHelper.java

```diff
@@ -36,8 +38,82 @@
  * @author Steve Ebersole
  */
 public class EnhancementHelper {
-	static void enhance(SourceSet sourceSet, EnhanceExtension options, Project project) {
-		final ClassLoader classLoader = toClassLoader( sourceSet.getRuntimeClasspath() );
+	public static void enhance(
+			DirectoryProperty classesDirectoryProperty,
+			InputChanges inputChanges,
+			EnhancementSpec enhancementDsl,
+			Project project) {
+		final Directory classesDir = classesDirectoryProperty.get();
+		final File classesDirFile = classesDir.getAsFile();
+
+		final Enhancer enhancer = generateEnhancer( classesDir, enhancementDsl );
+
+		final String classesDirPath = classesDirFile.getAbsolutePath();
+
+		inputChanges.getFileChanges( classesDirectoryProperty ).forEach(
+				change -> {
+					switch ( change.getChangeType() ) {
+						case ADDED:
+						case MODIFIED: {
+							final File changedFile = change.getFile();
+							if ( changedFile.getName().endsWith( ".class" ) ) {
+								final String classFilePath = changedFile.getAbsolutePath();
+								if ( classFilePath.startsWith( classesDirPath ) ) {
+									// we found the directory it came from
+									//		-use that to determine the class name
+									enhance( classesDirFile, changedFile, enhancer, project );
+									break;
+								}
+							}
+							break;
+						}
+						case REMOVED: {
+							// nothing to do
+							break;
+						}
+						default: {
+							throw new UnsupportedOperationException( "Unexpected ChangeType : " + change.getChangeType().name() );
+						}
+					}
+				}
+		);
+	}
+
+	private static void enhance(
+			File classesDir,
+			File javaClassFile,
+			Enhancer enhancer,
+			Project project) {
+		final byte[] enhancedBytecode = doEnhancement( classesDir, javaClassFile, enhancer );
+		if ( enhancedBytecode != null ) {
+			writeOutEnhancedClass( enhancedBytecode, javaClassFile, project.getLogger() );
```

Comment:
From what I can see, you're overwriting the original class file with the enhanced class file. Is that right?

If so, what's the point of the "output directory" property in `EnhancementSpec`?

Reply:
Initially I was thinking of a Javassist-like ability to write the corresponding Java source file for the enhancement.  But ByteBuddy does not seem to support that.  I'll remove it